File: 22-generate-parentheses/generate-parentheses.cpp
#include <vector>
#include <string>

class Solution {
public:
    std::vector<std::string> generateParenthesis(int n) {
        std::vector<std::string> res;
        std::string current;
        backtrack(res, current, 0, 0, n);
        return res;
    }
    
private:
    void backtrack(std::vector<std::string>& res, std::string& current, int open, int close, int n) {
        if (current.length() == 2 * n) {
            res.push_back(current);
            return;
        }
        if (open < n) {
            current.push_back('(');
            backtrack(res, current, open + 1, close, n);
            current.pop_back();
        }
        if (close < open) {
            current.push_back(')');
            backtrack(res, current, open, close + 1, n);
            current.pop_back();
        }
    }
};

File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []

        def cc(open,close,path):
            if len(path) == 2 * n:
                res.append("".join(path))
                return
            if open < n:
                path.append('(')
                cc(open + 1, close, path)
                path.pop()
            if close < open:
                path.append(')')
                cc(open , close + 1, path)
                path.pop()
        cc(0,0,[])
        return res